### ISSUE TITLE
Tools -> View script error uses variable width font, '^' for error position useless!

### ISSUE THREAD

**Issue by LorenAmelang** [autokey/autokey#206]
## Classification: Bug

(Pick one of: Bug, Crash/Hang/Data Loss, Performance, UI/Usability, Feature (New), Enhancement)

## Reproducibility: Always

(Pick one of: Always, Sometimes, Rarely, Unable, I Didn't Try)

## Summary
Tools -> View script error uses variable width font, '^' for error position useless!

## Steps to Reproduce (if applicable)
Create an error
Choose Tools -> View script error

## Expected Results
The '^' under the error statement should point to the problem

## Actual Results
Because both use variable width fonts, the position is random. 
You can't copy the error text to a fixed width editor. 
There doesn't seem to be any other log. 

If I've missed the solution, please provide a clue! 

## Version

AutoKey version: 0.95.3

Used GUI (Gtk, Qt, or both): Gtk

If the problem is known to be present in more than one version, please list all of those.

Installed via: (PPA, pip3, …). AUR

Distro: Antergos 18

**Comment by luziferius** [comment]
Thanks for pointing that out. The dialogue boxes should have a _Details_ area, rendered using a monospace font.
This can be easily implemented for autokey-qt, but I currently have no idea, if that is possible using the GTk interface without writing a whole message dialogue class…

In the meantime, you can see all `info` rated massages and all error messages in the autokey log file located at `~/.local/share/autokey/autokey.log`.
Alternatively, you can exit autokey and then run `autokey-gtk --verbose` in a terminal window. Then you can watch the output in real-time, including the error messages you want to see.

**Comment by LorenAmelang** [comment]
Appreciate the reply! I did find the log, and the verbose mode. And found my stupid error...  

I wanted to use the Qt version, but it failed here:
```
`[loren@Gazp9 ~]$ autokey-qt --verbose
Traceback (most recent call last):
  File "/usr/bin/autokey-qt", line 11, in <module>
    load_entry_point('autokey==0.95.3', 'console_scripts', 'autokey-qt')()

...

  File "/usr/lib/python3.7/site-packages/autokey/qtui/scriptpage.py", line 20, in <module>
    from PyQt5 import Qsci

ImportError: cannot import name 'Qsci' from 'PyQt5' (/usr/lib/python3.7/site-packages/PyQt5/__init__.py)

[loren@Gazp9 ~]$
`
```

With that I found a clue: 
--> You need to install qscintilla separately. If PyQt is already installed, then you should just have to install the qscintilla python bindings. 

And yes, only the bindings were missing. The Qt version is much nicer! 

Except some of the labels are cut off, even after I've drug the window over twice as big as it opens. 
![autokey truncated labels](https://user-images.githubusercontent.com/1297090/47957571-7ca29e80-df75-11e8-9e8a-445bd9939679.jpg)

BTW, that script runs without error, but doesn't actually do the bind, even if my sudo password is within its timeout. Guess I need to go searching...

**Comment by josephj11** [comment]
I'm a Python beginner so I may be wrong, but don't you have to split the arguments into separate strings? I do this with normal commands, but sudo is a bit trickier to call correctly, so I'm not sure.
Also, unless you specifically need bash (not apparent), why can't you just run sudo ...?

**Comment by LorenAmelang** [comment]
@josephj11
I'm probably more of a beginner than you! I've seen that "each argument is a separate string" style, and I think you're probably right. I just redid my whole project according to this:

[https://askubuntu.com/questions/155791/how-do-i-sudo-a-command-in-a-script-without-being-asked-for-a-password/155827#155827](url)

Rewritten for my goal:

Create a shell script with just that command
--> /home/loren/rfcommbind.sh
--> rfcomm bind 0 00:0A:3A:20:D4:C1
Quit the editor, and from the terminal, make the script executable and change its ownership to root, otherwise another user with access to your system could possibly edit it and execute whatever commands they want as root without needing your password:
sudo chown root:root /home/loren/rfcommbind.sh
sudo chmod 700 /home/loren/rfcommbind.sh

Set up sudo to allow rfcommbind.sh to execute without requiring a password
Type sudo visudo at the terminal to open the sudo permissions (sudoers) file
Near the end of the file, you'll see this line: %sudo   ALL=(ALL:ALL) ALL
Below that line, insert the following line, where username is your username:
loren  ALL=(ALL) NOPASSWD: /home/loren/rfcommbind.sh
Exit the editor (Ctrl+X if nano)

Modify your python script to call /home/loren/rfcommbind.sh
Change the line to:
os.system('sudo /home/loren/rfcommbind.sh')
or for AutoKey:
subprocess.run(['sudo', '/home/loren/rfcommbind.sh'])
Now your script should run without requiring a password AND without compromising the security of your account, your data or your system!

The python version works:
```
[loren@Gazp9 ~]$ rfcomm show /dev/rfcomm0
Get info failed: No such device
[loren@Gazp9 ~]$ python
Python 3.7.1 (default, Oct 22 2018, 10:41:28)
[GCC 8.2.1 20180831] on linux
>>> import os
>>> os.system('sudo /home/loren/rfcommbind.sh')
0
>>> quit()
[loren@Gazp9 ~]$ rfcomm show /dev/rfcomm0
rfcomm0: 00:0A:3A:20:D4:C1 channel 1 clean
[loren@Gazp9 ~]$
```
But in AutoKey, nothing appears to happen: 
```
2018-11-04 21:17:10,131 DEBUG - service - Raw key: 'h', modifiers: [<Key.SHIFT: '<shift>'>, <Key.SUPER: '<super>'>], Key: H
2018-11-04 21:17:10,132 DEBUG - service - Window visible title: 'loren@Gazp9:~', Window class: 'gnome-terminal-server.Gnome-terminal'
2018-11-04 21:17:10,132 INFO - service - Matched hotkey phrase/script with prompt=False
2018-11-04 21:17:10,132 DEBUG - service - Script runner executing: Script('TD40')
2018-11-04 21:17:10,136 ERROR - service - Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/autokey/service.py", line 207, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
[sudo] password for loren: 

```

Guess I need to search on "RuntimeError: release unlocked lock". Too late tonight...

**Comment by josephj11** [comment]
I can tell you all the Python I know in about five minutes. However, I'm pretty good with bash and the CLI. Let's break this into several issues.

First, the non-AutoKey ones.
rfcomm (which I've never seen before) does mess with the kernel, so it probably does need root access to work.

You'd probably be better off storing such a script somewhere safe  and not in your home directory where it's root ownership may get in the way. And that directory isn't in your PATH or in root's PATH. Something like that would usually be saved in somewhere like `/usr/local/bin` which is in both PATHs by default on many systems. However, since you used its full path, your command will find it anyway.

`os.system()` still works, but it is depreciated. The `subprocess` module has newer functions and is what you're supposed to use for all new code. I don't know the fine points, but it seems easier to use because it handles the arguments without your having to assemble the whole command as one string, white space and all.

The AutoKey message about the unlocked lock is a programming problem within AutoKey that occurs very frequently, but  doesn't cause any issues. You can ignore it.

That takes care of all the things that didn't make your command fail, but might bite you in the future. ;)

The part of the trace you included shows AutoKey detected Shift+H as a hotkey (fine if you never need to use a capital H for anything else ;) ) and then it ran a script named `TD40`. Since your script produces no visible effects, you'd have to run `rfcomm show /dev/rfcomm0` in a terminal (or do something similar) to see if it worked - assuming you reset it first so it wasn't "on" to begin with.

While we're on that subject, since most commands including `rfcomm` can fail, you probably need to add some error detection and handling to your script. Python has `try:` for that. Here's some [info](https://www.tutorialspoint.com/python3/python_exceptions.htm) on using it.

Which brings us (at last) to the real symptom that something didn't work.

The main problem appears to be that your `sudoers` file didn't work because `sudo` is shown asking for loren's password. It's expecting to get that on stdin, but that's not connected to your terminal when AutoKey is executing your script, so it just gets an end of file and fails.

I'm not sure why it works in interactive Python, but not in your AutoKey script. I assume you were logged in as loren when you ran it. AutoKey shouldn't do anything to alter the effective user.

Maybe someone else knows what that's about.

**Comment by LorenAmelang** [comment]
@josephj11
Thank you for saving me from the "AutoKey message about the unlocked lock" search! 

The rfcomm module is a maddeningly obsolete but apparently unavoidable link in connecting anything to a Bluetooth SPP protocol serial port - like the 20-year-old microcomputer that runs my off-grid house. 

Yes, I run rfcomm show /dev/rfcomm0 over and over, and release it between tries. Guess I should figure out how to put that in my AutoKey...

Today was about testing various options, and controlling the sudo password more carefully. I still have no idea when various parts of this system think I'm logged in as loren or sudo. SSH terminals seem to log me in when opened, but that doesn't log me into the notebook screen. And the notebook screen logs me out on a short timeout, even if I'm still logged in on SSH or VNC, unless it is hibernating. Then it can wake up logged in days later. 

I read a lot of web pages about subprocess...

--> 181105 In SSH term:
```
>>> subprocess.Popen(['sudo', '/home/loren/rfcommbind'])
<subprocess.Popen object at 0x7f459cc71fd0>
>>> sudo: /home/loren/rfcommbind: command not found
   [no prompt, pressed Enter]

>>> p = subprocess.Popen(['sudo', '/home/loren/rfcommbind'])
>>> sudo: /home/loren/rfcommbind: command not found
   [no prompt, pressed Enter]
>>> print (p)
<subprocess.Popen object at 0x7f459cc69eb8>
>>>
---
--> Or:
---
>>> subprocess.check_output(['sudo', '/home/loren/rfcommbind'])
sudo: /home/loren/rfcommbind: command not found
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/subprocess.py", line 389, in check_output
    **kwargs).stdout
  File "/usr/lib/python3.7/subprocess.py", line 481, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['sudo', '/home/loren/rfcommbind']' returned non-zero exit status 1.

>>> p = subprocess.check_output(['sudo', '/home/loren/rfcommbind'])
sudo: /home/loren/rfcommbind: command not found
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/subprocess.py", line 389, in check_output
    **kwargs).stdout
  File "/usr/lib/python3.7/subprocess.py", line 481, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['sudo', '/home/loren/rfcommbind']' returned non-zero exit status 1.
>>> print (p)
<subprocess.Popen object at 0x7f459cc71240>
>>>
```
If sudo password is active (during timeout), python console command works. 
If sudo is expired, it does not. 
No visible difference in console or log.
Neither condition works within AutoKey.

The AutoKey log doesn't show any more info:
```
2018-11-05 12:22:39,888 ERROR - service - Script error
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/autokey/service.py", line 467, in execute
    exec(script.code, scope)
  File "<string>", line 2, in <module>
  File "/usr/lib/python3.7/subprocess.py", line 389, in check_output
    **kwargs).stdout
  File "/usr/lib/python3.7/subprocess.py", line 481, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['sudo', '/home/loren/rfcommbind']' returned non-zero exit status 1.
```

All in one '' doesn't work:
`FileNotFoundError: [Errno 2] No such file or directory: 'sudo /home/loren/rfcommbind': 'sudo /home/loren/rfcommbind'
`

Separate '' '' don't work:
`subprocess.CalledProcessError: Command '['sudo', '/home/loren/rfcommbind']' returned non-zero exit status 1.
`

But if sudo password is live, this works:
```
>>> subprocess.run(['sudo', '/home/loren/rfcommbind.sh'])
CompletedProcess(args=['sudo', '/home/loren/rfcommbind.sh'], returncode=0)
```
I wish running autokey-qt --verbose would put the verbose error messages in the saved log. It includes tons of "DEBUG" stuff, but not my errors! They only show in the local console where the GUI was run.

**Comment by josephj11** [comment]
Some individual responses. This would be better on our [email list](https://groups.google.com/forum/#!forum/autokey-users) where I could reply inline.

You are never logged in as "sudo", it's not a user ID. There are some rules about what your effective user ID is when using `su` and `sudo` and what environment you are operating in when using them. I have read them several times, but still get confused. In your case, you're either loren or root on your main machine. Once you're on the other machine via `ssh`, your user there is probably root, but you can setup ssh  any way you want. I don't have experience with `ssh`.

Putting the `rfcomm` query into AutoKey first would be a good idea because AFAIK, it doesn't need root so that's a level of complexity simpler. The issue there is getting visible results of some sort. If I were coding it, I would probably save the command output in a string and display it using `dialog.info_dialog(title="Information", message="your-string")`. Or, since I'm a fan of `yad`, I'd probably use that instead via `subprocess`.

In `bash`, (you can get to it via `subprocess`), you can `echo $USER` to see who you are.
There's just the matter of login user vs. effective user which I'm not clear on.

I believe `ssh` logs you into a virtual terminal - no `X` or screen involved. There are ways to do a remote `X` session, but that's way beyond what we're dealing with here.

`/home/loren/rfcommbind` either does not exist (unlikely) or it doesn't have execute permissions for the effective user (which I thought you got right) to come back not found.

Your `subprocess` command returned an error and you have no `try:` exception handler, so it throws a traceback.

Your "root" problem (pun intended) appears to be that sudo isn't getting the memo that it's cool to run the command without password. That means one of two things is wrong: 1) The `sudoers` file is wrong or 2) the effective user when the command is run is not loren.

Neither of these are AutoKey problems, so we shouldn't really work on them here. On the support list, it would be fine.

Basically, you should learn to walk before you try to run. This particular situation involves Linux subtleties like getting `sudoers` right and effective user. Things like that are hard enough when you are already familiar with the basic tools like Python and the Linux CLI.

**Comment by LorenAmelang** [comment]
Commenting moved to:
[https://groups.google.com/forum/#!topic/autokey-users/gpQFjZHVGjk](url)

Apologies for not moving sooner! 

Loren

**Comment by LorenAmelang** [comment]
Well, lets try this again...  The discussion is moved to: 

Pasted without link button: 
https://groups.google.com/forum/#!topic/autokey-users/gpQFjZHVGjk
(Preview says Google...)

Pasted, selected, link button:
[](https://groups.google.com/forum/#!topic/autokey-users/gpQFjZHVGjk)
(Preview invisible)

Link button, then paste:
[https://groups.google.com/forum/#!topic/autokey-users/gpQFjZHVGjk](url)
Preview says Github...)

In Write, all of them now look like they point to Google Groups. But so did the previous try...  It still does, to the eye, but the link is to: https://github.com/autokey/autokey/issues/url

Wonder if any of these will work...

**Comment by LorenAmelang** [comment]
Joe clued me in to the "link" secret. you have paste it between the '[' and ']', AND between the '(' and')' (replacing "url".) 

[https://groups.google.com/forum/#!topic/autokey-users/gpQFjZHVGjk](https://groups.google.com/forum/#!topic/autokey-users/gpQFjZHVGjk)

**Comment by luziferius** [comment]
I know, this was a long time… I got some motivation and fixed it for the Qt GUI, so here is a small progress report. The dialog window will look like this in the next release:
![AK script error dialog](https://user-images.githubusercontent.com/2143820/68462970-fa932880-020d-11ea-9203-af8e5c595cf2.png)
The text can be selected and the window does not grow out of the screen bounds if the content gets long.


This still has to be done for the GTK GUI though …
